src/importer/sources/retroarch_source.py

```diff
@@ -0,0 +1,119 @@
+# retroarch_source.py
+#
+# Copyright 2023 Rilic
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from pathlib import Path
+from time import time
+
+import json
+import os
+import logging
+from json import JSONDecodeError
+
+from src import shared
+from src.game import Game
+from src.importer.sources.location import Location
+from src.importer.sources.source import (
+    SourceIterationResult,
+    SourceIterator,
+    Source
+)
+
+
+class RetroarchSourceIterator(SourceIterator):
+    source: "RetroarchSource"
+
+    def generator_builder(self) -> SourceIterationResult:
+        playlist_files = []
+        for file in os.listdir(self.source.config_location["playlists"]):
+            if file.endswith('.lpl'):
+                playlist_files.append(file)
+
+        playlist_items = []
+        for playlist_file in playlist_files:
+            open_file = open(str(self.source.config_location["playlists"]) + "/" + playlist_file)
+            try:
+                playlist_json = json.load(open_file)
+            except (JSONDecodeError, OSError, KeyError):
+                logging.warning("Cannot read playlist file: %s", str(playlist_file))
+                continue
+
+            for item in playlist_json["items"]:
+                # Select the core. Try the content's core first, then the playlist's
+                # default core.
+                core_path = item["core_path"]
```

Comment:
The idea is good, but the format can be improved. Also, the default core can also be `DETECT`.
What I suggest is something along those lines:

```py
for core_path in (
    item["core_path"],
    playlist_json["default_core_path"],
):
    if core_path != "DETECT":
        break
else:
    logging.warning("Cannot find core for: %s", str(item["path"]))
    continue
```

Reply:
Done this just now!
I think an error message popup showing all Retroarch games with no usable cores in the playlist would be useful for helping the user find out why some of their games are missing. Not sure how to implement that myself, though, so I'm leaving it as a log warning for now :sweat_smile: